Fix update bug of LaserMaterial

Mateasek opened the pull request:
Fixes problem described in #420 

In short:
Laser primitives are now dropped and recreated after Laser.transform changes which removes the problem of old transfomrs being cached in LaserMaterial

Mateasek committed:
Fix update bug of LaserMaterial

Laser primitives have to be rebuild with the materials.
Possible source of the problem:
Laser node is the parent of the laser primitives
After laser.transform is changed the _modified method is called before
the transforms of the laser primitives is changed because the Laser node
is above the laser primitives in the scenegraph.

Mateasek committed:
Update changelog

Mateasek committed:
Update changelog

vsnever reviewed (approved):
Hi @Mateasek, nice catch!

Ideally, if a material caches geometry of the primitive, any change in the geometry should automatically trigger cache rebuild. However, I don't know how to implement this, because laser segments are just basic primitives with an empty `_modified()` method. Considering this, your solution is fine.

vsnever commented:
Just as an option, wouldn't it be safer in general to cache affine matrices the first time the `emission_function()` is called, rather than when the `LaserMaterial` is initialised? In this case, the  `LaserMaterial` must keep the references to laser and laser segment nodes, but the `Laser` does not have to rebuild its internal geometry when the node is transformed.

Mateasek commented:
Hi @vsnever 

thanks for looking into it. The reaction to the material change would mean that the laser primitives couldn't be regular Raysect primitives, which I would like to avoid.

I like the idea of caching the transforms in the emission function. It is indeed the safest option. Thanks for suggesting it. I will apply the changes. I don't think that rebuilding the primitives on every change is a big deal, but from the point of view of safety your idea is much better.

Mateasek committed:
Change LaserMaterial transforms caching

The transforms are cached first time the emission method is called
This should be a safer option
The LaserMaterial now can be assigned only to a single primitive

Mateasek commented:
As suggested by @vsnever the transforms are now checked and cached in the emission function.

Also I added a check for the number of primitives the material is assigned to. Because of the transform caces the material should be assigned to only a single primitive for safety.

vsnever reviewed (approved):
Thanks @Mateasek, I like this fix better. Could you please merge the development branch into this, because it now has a fix for #422?

Mateasek committed:
Merge branch 'development' into fix/laser_material_update

Mateasek commented:
@vsnever, I merged the latest changes from development.

jacklovell reviewed (approved):
Looks OK to me. Thanks Matej!
